File: mobile/library/swift/src/GRPCClient.swift
import Foundation

/// Client that supports sending and receiving gRPC traffic.
@objcMembers
public final class GRPCClient: NSObject {
  private let httpClient: HTTPClient

  /// Create a new gRPC client instance.
  ///
  /// - parameter httpClient: The HTTP client to use for gRPC streams.
  public init(httpClient: HTTPClient) {
    self.httpClient = httpClient
  }

  /// Send a gRPC request with the provided handler.
  ///
  /// - parameter request: The outbound gRPC request. See `GRPCRequestBuilder` for creation.
  /// - parameter handler: Handler for receiving responses.
  ///
  /// - returns: An emitter that can be used for sending more traffic over the stream.
  public func send(_ request: Request, handler: GRPCResponseHandler) -> GRPCStreamEmitter {
    let emitter = self.httpClient.send(request, handler: handler.underlyingHandler)
    return GRPCStreamEmitter(emitter: emitter)
  }
}

File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/GRPCClient.kt
package io.envoyproxy.envoymobile

// 1 byte for the compression flag, 4 bytes for the message length (int)
const val GRPC_PREFIX_LENGTH = 5

/**
 * Client that supports sending and receiving gRPC traffic.
 *
 * @param httpClient The HTTP client to use for gRPC streams.
 */
class GRPCClient(
    private val httpClient: HTTPClient
) {

  /**
   * Start a gRPC request with the provided handler.
   *
   * @param request The outbound gRPC request. See `GRPCRequestBuilder` for creation.
   * @param handler Handler for receiving responses.
   *
   * @returns GRPCStreamEmitter, An emitter that can be used for sending more traffic over the stream.
   */
  fun start(request: Request, grpcResponseHandler: GRPCResponseHandler): GRPCStreamEmitter {
    val emitter = httpClient.start(request, grpcResponseHandler.underlyingHandler)
    return GRPCStreamEmitter(emitter)
  }
}
